File: core/src/main/java/com/ataulm/chunks/Chunks.java
package com.ataulm.chunks;

import com.ataulm.Optional;
import com.google.auto.value.AutoValue;

import javax.annotation.Nullable;

@AutoValue
public abstract class Chunks {

    public static Chunks empty(ChunkDate todaysDate) {
        return create(todaysDate, Items.empty(), Items.empty(), Items.empty());
    }

    public static Chunks create(ChunkDate todaysDate, Items today, Items tomorrow, Items sometime) {
        return create(todaysDate, today, tomorrow, sometime, null);
    }

    public static Chunks create(ChunkDate todaysDate, Items today, Items tomorrow, Items sometime, @Nullable String input) {
        return new AutoValue_Chunks(todaysDate, today, tomorrow, sometime, Optional.fromNullable(input));
    }

    protected Chunks() {
        // use static factory
    }

    public abstract ChunkDate todaysDate();

    public abstract Items today();

    public abstract Items tomorrow();

    public abstract Items sometime();

    public abstract Optional<String> input();

}

File: core/src/main/java/com/ataulm/chunks/Chunks.kt
package com.ataulm.chunks

data class Chunks(
        val todaysDate: ChunkDate,
        val today: Items,
        val tomorrow: Items,
        val sometime: Items,
        val input: String?
) {

    companion object {

        fun empty(todaysDate: ChunkDate): Chunks {
            return create(todaysDate, Items.empty(), Items.empty(), Items.empty())
        }

        @JvmOverloads
        fun create(todaysDate: ChunkDate, today: Items, tomorrow: Items, sometime: Items, input: String? = null): Chunks {
            return Chunks(todaysDate, today, tomorrow, sometime, input)
        }
    }
}

